### PR TITLE
Core: unify map conversions

### PR DESCRIPTION
**Pull request by acarbonetto**
based off of: https://github.com/aws/glide-for-redis/compare/main...Bit-Quill:glide-for-redis:java/dbg_yuryf_xread

### CODE REVIEW

File: glide-core/src/client/value_conversion.rs [hunk 1]
```diff
@@ -6,8 +6,11 @@ use redis::{
 };
 
 #[derive(Clone, Copy)]
-pub(crate) enum ExpectedReturnType {
-    Map,
+pub(crate) enum ExpectedReturnType<'a> {
+    Map(
+        &'a Option<ExpectedReturnType<'a>>,
+        &'a Option<ExpectedReturnType<'a>>,
```
Comment (Yury-Fridlyand) [inline glide-core/src/client/value_conversion.rs:12]:
```suggestion
        &'a Option<ExpectedReturnType<'a>>, // map key type
        &'a Option<ExpectedReturnType<'a>>, // map valye type
```
I assume you can omit the key type - it is always a string

File: glide-core/src/client/value_conversion.rs [hunk 2]
```diff
@@ -50,34 +53,12 @@ pub(crate) fn convert_to_expected_type(
         },
         ExpectedReturnType::MapOfStringToDouble => match value {
```
Comment (Yury-Fridlyand) [inline glide-core/src/client/value_conversion.rs:54]: you can replace `MapOfStringToDouble` with `Map(&Some(BulkString), &Some(Double))` and delete it
Same with other `MapOf...` types we have

File: glide-core/src/client/value_conversion.rs [hunk 3]
```diff
@@ -411,7 +392,26 @@ fn convert_array_elements(
     Ok(Value::Array(converted_array))
 }
 
-fn convert_array_to_map(
+fn convert_inner_map_by_type(
```
Comment (Yury-Fridlyand) [inline glide-core/src/client/value_conversion.rs:395]: Doc for the god of docs!

File: glide-core/src/client/value_conversion.rs [hunk 4]
```diff
@@ -527,22 +527,34 @@ pub(crate) fn expected_type_for_cmd(cmd: &Cmd) -> Option<ExpectedReturnType> {
...
+            &Some(ExpectedReturnType::BulkString),
+            &Some(ExpectedReturnType::Double)
+        )),
         b"BZMPOP" | b"ZMPOP" => Some(ExpectedReturnType::ZMPopReturnType),
         b"JSON.TOGGLE" => Some(ExpectedReturnType::JsonToggleReturnType),
         b"GEOPOS" => Some(ExpectedReturnType::ArrayOfArraysOfDoubleOrNull),
-        b"LMPOP" => Some(ExpectedReturnType::ArrayOfStringAndArrays),
-        b"BLMPOP" => Some(ExpectedReturnType::ArrayOfStringAndArrays),
+        b"LMPOP" | b"BLMPOP" => Some(ExpectedReturnType::Map(
+            &Some(ExpectedReturnType::BulkString),
```
Comment (Yury-Fridlyand) [inline glide-core/src/client/value_conversion.rs:555]: Delete unused types

File: glide-core/src/client/value_conversion.rs (hunk 1 above)
Comment (jonathanl-bq) [reply glide-core/src/client/value_conversion.rs:12]: I'd prefer using curly brace syntax here to give names to the fields of the enum over just adding comments. Make the code a bit more self-documenting.
```suggestion
    Map {
        key_type: &'a Option<ExpectedReturnType<'a>>,
        value_type: &'a Option<ExpectedReturnType<'a>>,
    }
```

**Review by aaron-congo — APPROVED** [review]
We should maybe add docs for convert_inner_map_by_type and convert_array_to_map_by_type like Yury suggested but otherwise LGTM

File: glide-core/src/client/value_conversion.rs [hunk 5]
```diff
@@ -411,7 +348,25 @@ fn convert_array_elements(
     Ok(Value::Array(converted_array))
 }
 
-fn convert_array_to_map(
+fn convert_inner_map_by_type(
```
Comment (aaron-congo) [inline glide-core/src/client/value_conversion.rs:351]: What do you mean by "inner_map"? It looks like this function converts map keys and values, and those values may or may not also be maps. Would "convert_map_by_type" or "convert_map_entries" be clearer?

**Pull request closed without merging**